src/hotspot/share/opto/predicates.cpp

```diff
@@ -77,13 +77,29 @@ Deoptimization::DeoptReason RuntimePredicate::uncommon_trap_reason(IfProjNode* i
 }
 
 bool RuntimePredicate::is_success_proj(Node* node, Deoptimization::DeoptReason deopt_reason) {
-  if (node->is_IfProj() && !node->in(0)->is_ParsePredicate()) {
+  if (may_be_runtime_predicate_if(node)) {
     return deopt_reason == uncommon_trap_reason(node->as_IfProj());
   } else {
     return false;
   }
 }
 
+// A Runtime Predicate must have an If or a RangeCheck node, while the If should not be a zero trip guard check.
+bool RuntimePredicate::may_be_runtime_predicate_if(Node* node) {
+  if (node->is_IfProj()) {
+    const int opcode_if = node->in(0)->Opcode();
+    if ((opcode_if == Op_If && is_not_zero_trip_guard(node->as_IfProj()))
+        || opcode_if == Op_RangeCheck) {
+      return true;
+    }
+  }
+  return false;
+}
+
+bool RuntimePredicate::is_not_zero_trip_guard(const IfProjNode* if_proj) {
```

Comment:
I don't think it's worth adding a new method for this, especially since we should prefer negation of the result in the caller instead of using negation in method/variable names.